File: java/BaseAdapter.java
package com.lakeraven.filebot.core.adapters;

import java.util.concurrent.CompletableFuture;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.time.Instant;
import com.lakeraven.filebot.core.models.*;

/**
 * Abstract base class for all MUMPS database adapters
 * 
 * This interface ensures implementation consistency across different MUMPS platforms
 * (InterSystems IRIS, YottaDB, GT.M) while enabling high-performance healthcare operations.
 */
public abstract class BaseAdapter {
    
    protected Configuration config;
    protected boolean connected = false;
    
    /**
     * Constructor with configuration
     * @param config Adapter-specific configuration
     */
    public BaseAdapter(Configuration config) {
        this.config = config;
        setupConnection();
    }
    
    // ==================== Core Global Operations ====================
    
    /**
     * Get value from global node
     * @param global Global name (e.g., "^DPT")
     * @param subscripts Variable number of subscripts
     * @return Global value or null if not set
     */
    public abstract CompletableFuture<String> getGlobal(String global, String... subscripts);
    
    /**
     * Set value in global node
     * @param value Value to set
     * @param global Global name
     * @param subscripts Variable number of subscripts
     * @return Success status
     */
    public abstract CompletableFuture<Boolean> setGlobal(String value, String global, String... subscripts);
    
    /**
     * Get next subscript in order
     * @param global Global name
     * @param subscripts Current subscripts
     * @return Next subscript or null if no more
     */
    public abstract CompletableFuture<String> orderGlobal(String global, String... subscripts);
    
    /**
     * Check if global node has data (defined)
     * @param global Global name
     * @param subscripts Variable number of subscripts
     * @return 0=undefined, 1=data, 10=descendants, 11=both
     */
    public abstract CompletableFuture<Integer> dataGlobal(String global, String... subscripts);
    
    // ==================== Adapter Identification ====================
    
    /**
     * Get adapter type identifier
     * @return Adapter type ("iris", "yottadb", "gtm", etc.)
     */
    public abstract String getAdapterType();
    
    /**
     * Check if adapter is connected and ready
     * @return Connection status
     */
    public boolean isConnected() {
        return connected;
    }
    
    /**
     * Get adapter version information
     * @return Version info with adapter and database versions
     */
    public VersionInfo getVersionInfo() {
        return new VersionInfo("1.0.0", "unknown");
    }
    
    /**
     * Get adapter capabilities
     * @return Capabilities with boolean flags
     */
    public Capabilities getCapabilities() {
        return new Capabilities(false, false, false, true, true, false);
    }
    
    // ==================== Connection Management ====================
    
    /**
     * Test adapter connectivity with database
     * @return Test result with success status and message
     */
    public CompletableFuture<ConnectionResult> testConnection() {
        if (!isConnected()) {
            return CompletableFuture.completedFuture(
                ConnectionResult.builder()
                    .success(false)
                    .message("Adapter not connected")
                    .build());
        }
        
        return CompletableFuture.supplyAsync(() -> {
            try {
                // Try a simple global operation
                String testGlobal = "^FILEBOT_TEST_" + System.currentTimeMillis();
                setGlobal("test", testGlobal, "connection").join();
                String result = getGlobal(testGlobal, "connection").join();
                setGlobal("", testGlobal, "connection").join(); // Cleanup
                
                if ("test".equals(result)) {
                    return ConnectionResult.builder()
                        .success(true)
                        .message("Connection successful")
                        .build();
                } else {
                    return ConnectionResult.builder()
                        .success(false)
                        .message("Global operation test failed")
                        .build();
                }
            } catch (Exception e) {
                return ConnectionResult.builder()
                    .success(false)
                    .message("Connection test failed: " + e.getMessage())
                    .build();
            }
        });
    }
    
    /**
     * Close adapter connection and cleanup resources
     * @return Cleanup success status
     */
    public CompletableFuture<Boolean> close() {
        connected = false;
        return CompletableFuture.completedFuture(true);
    }
    
    // ==================== Advanced Operations (Optional) ====================
    
    /**
     * Execute MUMPS code directly (optional for advanced adapters)
     * @param code MUMPS code to execute
     * @return Execution result
     * @throws UnsupportedOperationException if not supported
     */
    public CompletableFuture<String> executeMumps(String code) {
        return CompletableFuture.failedFuture(
            new UnsupportedOperationException(getAdapterType() + " adapter does not support MUMPS execution"));
    }
    
    /**
     * Lock global node (optional for locking-capable adapters)
     * @param global Global name
     * @param subscripts Subscripts to lock
     * @param timeout Lock timeout in seconds
     * @return Lock acquired successfully
     */
    public CompletableFuture<Boolean> lockGlobal(String global, List<String> subscripts, int timeout) {
        // Default implementation returns true (no-op)
        return CompletableFuture.completedFuture(true);
    }
    
    /**
     * Unlock global node (optional for locking-capable adapters)
     * @param global Global name
     * @param subscripts Subscripts to unlock
     * @return Unlock successful
     */
    public CompletableFuture<Boolean> unlockGlobal(String global, List<String> subscripts) {
        // Default implementation returns true (no-op)
        return CompletableFuture.completedFuture(true);
    }
    
    // ==================== Transaction Support (Optional) ====================
    
    /**
     * Start transaction (optional for transaction-capable adapters)
     * @return Transaction handle
     */
    public CompletableFuture<Transaction> startTransaction() {
        // Default implementation returns null (no transactions)
        return CompletableFuture.completedFuture(null);
    }
    
    /**
     * Commit transaction (optional for transaction-capable adapters)
     * @param transaction Transaction handle
     * @return Commit successful
     */
    public CompletableFuture<Boolean> commitTransaction(Transaction transaction) {
        // Default implementation returns true (no-op)
        return CompletableFuture.completedFuture(true);
    }
    
    /**
     * Rollback transaction (optional for transaction-capable adapters)
     * @param transaction Transaction handle
     * @return Rollback successful
     */
    public CompletableFuture<Boolean> rollbackTransaction(Transaction transaction) {
        // Default implementation returns true (no-op)
        return CompletableFuture.completedFuture(true);
    }
    
    // ==================== Protected Helper Methods ====================
    
    /**
     * Hook for adapter-specific setup (called during initialization)
     * Override in concrete adapters for custom setup logic
     */
    protected void setupConnection() {
        // Default implementation is no-op
    }
    
    /**
     * Validate subscripts for global operations
     * @param subscripts Subscripts to validate
     * @return Validated subscripts as strings
     */
    protected String[] validateSubscripts(String[] subscripts) {
        if (subscripts == null) {
            return new String[0];
        }
        
        String[] validated = new String[subscripts.length];
        for (int i = 0; i < subscripts.length; i++) {
            validated[i] = subscripts[i] != null ? subscripts[i] : "";
        }
        return validated;
    }
    
    /**
     * Normalize global name (ensure proper format)
     * @param global Global name to normalize
     * @return Normalized global name
     */
    protected String normalizeGlobalName(String global) {
        if (global == null || global.isEmpty()) {
            throw new IllegalArgumentException("Global name cannot be null or empty");
        }
        return global.startsWith("^") ? global : "^" + global;
    }
    
    /**
     * Validate adapter configuration
     * @return List of validation errors (empty if valid)
     */
    public List<String> validateConfig() {
        List<String> errors = new ArrayList<>();
        if (config == null) {
            errors.add("Configuration cannot be null");
        }
        return errors;
    }
}

File: python/base_adapter.py
"""
Abstract base adapter for FileBot Python implementation

This module defines the abstract base class that all MUMPS database adapters
must implement to ensure consistency and compatibility.
"""

from abc import ABC, abstractmethod
from typing import Optional, List, Dict, Any, Union
import asyncio
from dataclasses import dataclass
from datetime import datetime
import logging


@dataclass
class VersionInfo:
    """Version information for adapter and database"""
    adapter_version: str
    database_version: str


@dataclass
class Capabilities:
    """Adapter capabilities flags"""
    transactions: bool = False
    locking: bool = False
    mumps_execution: bool = False
    concurrent_access: bool = True
    cross_references: bool = True
    unicode_support: bool = False


@dataclass
class ConnectionResult:
    """Result of connection test"""
    success: bool
    message: str
    details: Optional[Dict[str, Any]] = None
    timestamp: Optional[datetime] = None


class Transaction:
    """Base transaction interface"""
    
    def __init__(self, handle: Any):
        self.handle = handle
        self.started_at = datetime.now()
        self.completed = False
    
    def get_handle(self) -> Any:
        return self.handle
    
    def is_completed(self) -> bool:
        return self.completed
    
    def mark_completed(self) -> None:
        self.completed = True


class BaseAdapter(ABC):
    """
    Abstract base adapter that defines the contract for all MUMPS database adapters
    
    This interface ensures implementation consistency across different MUMPS platforms
    (InterSystems IRIS, YottaDB, GT.M) while enabling high-performance healthcare operations.
    """
    
    def __init__(self, config: Dict[str, Any]):
        """
        Initialize adapter with configuration
        
        Args:
            config: Adapter-specific configuration parameters
        """
        self.config = config.copy() if config else {}
        self.connected = False
        self.logger = logging.getLogger(f"filebot.{self.__class__.__name__}")
        
        # Perform setup
        asyncio.create_task(self._setup_connection())
    
    # ==================== Core Global Operations ====================
    
    @abstractmethod
    async def get_global(self, global_name: str, *subscripts: str) -> Optional[str]:
        """
        Get value from global node
        
        Args:
            global_name: Global name (e.g., "^DPT")
            subscripts: Variable number of subscripts
            
        Returns:
            Global value or None if not set
        """
        pass
    
    @abstractmethod
    async def set_global(self, value: str, global_name: str, *subscripts: str) -> bool:
        """
        Set value in global node
        
        Args:
            value: Value to set
            global_name: Global name
            subscripts: Variable number of subscripts
            
        Returns:
            Success status
        """
        pass
    
    @abstractmethod
    async def order_global(self, global_name: str, *subscripts: str) -> Optional[str]:
        """
        Get next subscript in order
        
        Args:
            global_name: Global name
            subscripts: Current subscripts
            
        Returns:
            Next subscript or None if no more
        """
        pass
    
    @abstractmethod
    async def data_global(self, global_name: str, *subscripts: str) -> int:
        """
        Check if global node has data (defined)
        
        Args:
            global_name: Global name
            subscripts: Variable number of subscripts
            
        Returns:
            0=undefined, 1=data, 10=descendants, 11=both
        """
        pass
    
    # ==================== Adapter Identification ====================
    
    @abstractmethod
    def get_adapter_type(self) -> str:
        """
        Get adapter type identifier
        
        Returns:
            Adapter type (e.g., "iris", "yottadb", "gtm")
        """
        pass
    
    def is_connected(self) -> bool:
        """
        Check if adapter is connected and ready
        
        Returns:
            Connection status
        """
        return self.connected
    
    def get_version_info(self) -> VersionInfo:
        """
        Get adapter version information
        
        Returns:
            Version info with adapter and database versions
        """
        return VersionInfo("1.0.0", "unknown")
    
    def get_capabilities(self) -> Capabilities:
        """
        Get adapter capabilities
        
        Returns:
            Capabilities with boolean flags
        """
        return Capabilities()
    
    # ==================== Connection Management ====================
    
    async def test_connection(self) -> ConnectionResult:
        """
        Test adapter connectivity with database
        
        Returns:
            Test result with success status and message
        """
        if not self.is_connected():
            return ConnectionResult(False, "Adapter not connected")
        
        try:
            # Try a simple global operation
            test_global = f"^FILEBOT_TEST_{int(datetime.now().timestamp())}"
            await self.set_global("test", test_global, "connection")
            result = await self.get_global(test_global, "connection")
            await self.set_global("", test_global, "connection")  # Cleanup
            
            if result == "test":
                return ConnectionResult(True, "Connection successful")
            else:
                return ConnectionResult(False, "Global operation test failed")
                
        except Exception as e:
            return ConnectionResult(False, f"Connection test failed: {str(e)}")
    
    async def close(self) -> bool:
        """
        Close adapter connection and cleanup resources
        
        Returns:
            Cleanup success status
        """
        self.connected = False
        return True
    
    # ==================== Advanced Operations (Optional) ====================
    
    async def execute_mumps(self, code: str) -> str:
        """
        Execute MUMPS code directly (optional for advanced adapters)
        
        Args:
            code: MUMPS code to execute
            
        Returns:
            Execution result
            
        Raises:
            NotImplementedError: If not supported by adapter
        """
        raise NotImplementedError(f"{self.get_adapter_type()} adapter does not support MUMPS execution")
    
    async def lock_global(self, global_name: str, subscripts: List[str], timeout: int = 30) -> bool:
        """
        Lock global node (optional for locking-capable adapters)
        
        Args:
            global_name: Global name
            subscripts: Subscripts to lock
            timeout: Lock timeout in seconds
            
        Returns:
            Lock acquired successfully
        """
        # Default implementation returns True (no-op)
        return True
    
    async def unlock_global(self, global_name: str, subscripts: List[str]) -> bool:
        """
        Unlock global node (optional for locking-capable adapters)
        
        Args:
            global_name: Global name
            subscripts: Subscripts to unlock
            
        Returns:
            Unlock successful
        """
        # Default implementation returns True (no-op)
        return True
    
    # ==================== Transaction Support (Optional) ====================
    
    async def start_transaction(self) -> Optional[Transaction]:
        """
        Start transaction (optional for transaction-capable adapters)
        
        Returns:
            Transaction handle or None
        """
        # Default implementation returns None (no transactions)
        return None
    
    async def commit_transaction(self, transaction: Transaction) -> bool:
        """
        Commit transaction (optional for transaction-capable adapters)
        
        Args:
            transaction: Transaction handle
            
        Returns:
            Commit successful
        """
        # Default implementation returns True (no-op)
        if transaction:
            transaction.mark_completed()
        return True
    
    async def rollback_transaction(self, transaction: Transaction) -> bool:
        """
        Rollback transaction (optional for transaction-capable adapters)
        
        Args:
            transaction: Transaction handle
            
        Returns:
            Rollback successful
        """
        # Default implementation returns True (no-op)
        if transaction:
            transaction.mark_completed()
        return True
    
    # ==================== Protected Helper Methods ====================
    
    async def _setup_connection(self) -> None:
        """
        Hook for adapter-specific setup (called during initialization)
        Override in concrete adapters for custom setup logic
        """
        # Default implementation is no-op
        pass
    
    def _validate_subscripts(self, subscripts: tuple) -> List[str]:
        """
        Validate subscripts for global operations
        
        Args:
            subscripts: Subscripts to validate
            
        Returns:
            Validated subscripts as strings
        """
        return [str(sub) if sub is not None else "" for sub in subscripts]
    
    def _normalize_global_name(self, global_name: str) -> str:
        """
        Normalize global name (ensure proper format)
        
        Args:
            global_name: Global name to normalize
            
        Returns:
            Normalized global name
            
        Raises:
            ValueError: If global name is invalid
        """
        if not global_name:
            raise ValueError("Global name cannot be empty")
        
        return global_name if global_name.startswith("^") else f"^{global_name}"
    
    def validate_config(self) -> List[str]:
        """
        Validate adapter configuration
        
        Returns:
            List of validation errors (empty if valid)
        """
        errors = []
        
        if self.config is None:
            errors.append("Configuration cannot be None")
        
        return errors


# ==================== Adapter Registry Interface ====================

class AdapterRegistry:
    """Registry for managing MUMPS database adapters"""
    
    _instance = None
    _adapters = {}
    
    @classmethod
    def get_instance(cls):
        """Get singleton instance of adapter registry"""
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance
    
    def register_adapter(self, name: str, adapter_class: type, 
                        description: str = "", version: str = "1.0.0", 
                        priority: int = 0, auto_detect: bool = False) -> None:
        """
        Register an adapter
        
        Args:
            name: Adapter identifier
            adapter_class: Adapter class implementing BaseAdapter
            description: Adapter description
            version: Adapter version
            priority: Priority for auto-detection (higher = preferred)
            auto_detect: Include in auto-detection
        """
        self._adapters[name] = {
            'class': adapter_class,
            'description': description,
            'version': version,
            'priority': priority,
            'auto_detect': auto_detect
        }
    
    def get_adapter_class(self, name: str) -> Optional[type]:
        """
        Get adapter class by name
        
        Args:
            name: Adapter identifier
            
        Returns:
            Adapter class or None if not found
        """
        adapter_info = self._adapters.get(name)
        return adapter_info['class'] if adapter_info else None
    
    def create_adapter(self, name: str, config: Dict[str, Any]) -> BaseAdapter:
        """
        Create adapter instance
        
        Args:
            name: Adapter identifier
            config: Configuration for adapter
            
        Returns:
            Configured adapter instance
            
        Raises:
            ValueError: If adapter not found
        """
        if name == "auto_detect":
            name = self._auto_detect_adapter()
        
        adapter_class = self.get_adapter_class(name)
        if not adapter_class:
            available = list(self._adapters.keys())
            raise ValueError(f"Unknown adapter: {name}. Available: {available}")
        
        return adapter_class(config)
    
    def list_adapters(self) -> List[Dict[str, Any]]:
        """
        List all available adapters
        
        Returns:
            List of adapter information
        """
        return [
            {
                'name': name,
                'description': info['description'],
                'version': info['version'],
                'priority': info['priority']
            }
            for name, info in self._adapters.items()
        ]
    
    def _auto_detect_adapter(self) -> str:
        """
        Auto-detect best available adapter
        
        Returns:
            Name of best available adapter
            
        Raises:
            RuntimeError: If no adapters available
        """
        # Sort by priority, highest first
        available = [
            (name, info) for name, info in self._adapters.items()
            if info['auto_detect']
        ]
        
        if not available:
            raise RuntimeError("No adapters available for auto-detection")
        
        # Return highest priority adapter
        available.sort(key=lambda x: x[1]['priority'], reverse=True)
        return available[0][0]


# ==================== Example Implementation ====================

class MockAdapter(BaseAdapter):
    """Mock adapter for testing purposes"""
    
    def __init__(self, config: Dict[str, Any]):
        super().__init__(config)
        self.data = {}
        self.connected = True
    
    def get_adapter_type(self) -> str:
        return "mock"
    
    async def get_global(self, global_name: str, *subscripts: str) -> Optional[str]:
        key = self._build_key(global_name, *subscripts)
        return self.data.get(key)
    
    async def set_global(self, value: str, global_name: str, *subscripts: str) -> bool:
        key = self._build_key(global_name, *subscripts)
        if value:
            self.data[key] = value
        else:
            self.data.pop(key, None)
        return True
    
    async def order_global(self, global_name: str, *subscripts: str) -> Optional[str]:
        # Simple mock implementation
        key_prefix = self._build_key(global_name, *subscripts)
        matching_keys = [k for k in self.data.keys() if k.startswith(key_prefix)]
        return matching_keys[0] if matching_keys else None
    
    async def data_global(self, global_name: str, *subscripts: str) -> int:
        key = self._build_key(global_name, *subscripts)
        return 1 if key in self.data else 0
    
    def _build_key(self, global_name: str, *subscripts: str) -> str:
        normalized = self._normalize_global_name(global_name)
        if subscripts:
            return f"{normalized}({','.join(subscripts)})"
        return normalized


# Register mock adapter for testing
registry = AdapterRegistry.get_instance()
registry.register_adapter(
    "mock", 
    MockAdapter, 
    "Mock adapter for testing", 
    "1.0.0", 
    priority=1000,  # Highest priority for testing
    auto_detect=True
)